src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/inspectors/RepairPatch.java

```diff
@@ -4,135 +4,263 @@
+				&& Objects.equals(diff, that.diff);
+	}
+
+	@Override
+	public int hashCode() {
+
+		return Objects.hash(toolname, filePath, diff);
+	}
+
+	// ranking algorithms
+	public static Comparator<RepairPatch> rankByOverfittingWithFeatures(Features features) {
+		return (x, y) -> overfittingSort(x, y, features);
+	}
+
+	private static int overfittingSort(RepairPatch patch1, RepairPatch patch2, Features features) { // ascending
+		double score1 = patch1.getOverfittingScore(features);
+		double score2 = patch2.getOverfittingScore(features);
+		double diff = score1 - score2;
+		if (diff < 0) {
+			return -1;
+		} else if (diff > 0) {
+			return 1;
+		}
+		return 0;
+	}
+
+	// ODS classification
+	public static List<RepairPatch> classifyByODSWithFeatures(List<RepairPatch> allPatches, Long buildId) {
+
+		for (int patchID = 0; patchID < allPatches.size(); patchID++) {
+			RepairPatch repairPatch = allPatches.get(patchID);
+			String label  = repairPatch.computeODSLabel(patchID, buildId);
+			repairPatch.setODSLabel(label);
+		}
+		return allPatches;
+	}
+
+	private String computeODSLabel(int patchId, Long buildId) {
+		File buggyFile = new File(filePath);
+		// if no buggy file available, we provide the unknown label for the patches.
+		String label = "unknown";
+		if (!buggyFile.isFile()) {
+			return label;
+		}
+
+		// read from buggyFile
+		List<String> buggyLines = new ArrayList<>();
+		try (Stream<String> stream = Files.lines(Paths.get(filePath))) {
+			stream.forEach(buggyLines::add);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+
+		// prepare patches
+		List<String> diffLines = Arrays.asList(diff.split("\n"));
+		Patch<String> patches = UnifiedDiffUtils.parseUnifiedDiff(diffLines);
+
+		try {
+			// create a directory to store the patch: "patches/"+buildId+patchId
+			String buggyClassName = buggyFile.getName().split(".java")[0];
+			String rootfile = filePath.split("patch")[0] + "/patches";
+			Path path = Paths
+					.get(filePath.split("patch")[0] + "/patches/" + buildId + "-" + patchId + '/' + buggyClassName);
+			Files.createDirectories(path);
+
+			// create buggy file and patchedFile that follows Coming structure
+			File newbuggyFile = new File(path + "/" + buildId + "-" + patchId + "_" + buggyClassName + "_s.java");
+			File patchedFile = new File(path + "/" + buildId + "-" + patchId + "_" + buggyClassName + "_t.java");
+
+			// copy the buggy file under the patch folder
+			Files.write(Paths.get(newbuggyFile.getPath()), buggyLines);
+			// generate content of patchedFile by applying patches
+			List<String> patchedLines = DiffUtils.patch(buggyLines, patches);
+			Files.write(Paths.get(patchedFile.getPath()), patchedLines);
+
+			
+			label = extractODSFeatures(rootfile);
+						
+
+		} catch (PatchFailedException | IOException e) {
+			throw new RuntimeException(e);
+		}
+
+		return label;
+
+	}
+
+	private String extractODSFeatures(String rootfile) {
+//		try {
```

Comment:
This will be the most challenging part: 
1) call coming to compute features
2) summarize the 3 set of features
3) call python

I wonder if step 2 would be directly  included/implemented in Coming

Reply:
Hi Matias! Fully agree with you. That's what I was thinking!  We can directly output the feature values from coming. I will make add some code on the coming repo first.

Reply:
Perfect, let me know if you need some help with any task :)